### ISSUE TITLE
Crash on Android 10

### ISSUE THREAD

**Issue by RoProducts** [ant-wireless/ANT-Android-SDKs#14]
Using the pluginLib_3-6-0.jar on Android 10 results in the following crash : 

    java.lang.SecurityException: MODE_WORLD_READABLE no longer supported
      at android.app.ContextImpl.checkMode (ContextImpl.java:2788)
      at android.app.ContextImpl.getSharedPreferences (ContextImpl.java:471)
      at android.app.ContextImpl.getSharedPreferences (ContextImpl.java:461)
      at com.dsi.ant.plugins.utility.log.LogAnt.getDebugLevel (LogAnt.java:62)
      at com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc.bindAndRequest (AntPluginPcc.java:599)
      at com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc.requestAccess_Helper_SubMain (AntPluginPcc.java:568)
      at com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc.requestAccess_Helper_Main (AntPluginPcc.java:512)
      at com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc.requestAccess_Helper_AsyncSearchByDevNumber (AntPluginPcc.java:428)

on a Samsung Galaxy S10. Please fix as otherwise Ant+ cannot be used on Android 10.

Thank you

Robert

**Comment by iforpowell** [comment]
I have seen a crash report with this.  I don't know what happens if I build targeting a lower API level I don't have a suitable device to test on at present.  It's going to be getting the debug logging level from the ANT Services Settings App.  Maybe it will actually work if the app is not installed.  I would expect the code to deal with this case.

**Comment by harrychin** [comment]
This is a known compatibility issue which we are planning to address soon in a new API release. In the meantime, catching the "SecurityException" thrown by "bindAndRequest" should also be a suitable workaround.

**Comment by RoProducts** [comment]
Catching the exception is not a workaround as the device is not connected.

Please provide a fix asap.

Thank you

**Comment by harrychin** [comment]
We've posted a release candidate branch with a new Plugin Library which you should be able to easily replace your existing JAR with here:

https://github.com/ant-wireless/ANT-Android-SDKs/tree/development/C.B6-P.B6RC/ANT%2B_Android_SDK

Let us know how your experience with it goes.

**Comment by RoProducts** [comment]
Works out for me.

Thank you for your quick reaction.

**Comment by hgoebl** [comment]
Don't know whether it works on Android 10, but the lib seams to work on older devices, at least.

**Comment by harrychin** [comment]
> I have seen a crash report with this. I don't know what happens if I build targeting a lower API level I don't have a suitable device to test on at present. It's going to be getting the debug logging level from the ANT Services Settings App. Maybe it will actually work if the app is not installed. I would expect the code to deal with this case.

You should be able to replicate the crash by installing the 3.8.0 version of the AntPlusPluginService which now targets API level 29, as it was mandated by Google to target newer API levels which trigger the new exception.

**Comment by hgoebl** [comment]
> You should be able to replicate the crash by installing the 3.8.0 version of the AntPlusPluginService which now targets API level 29, as it was mandated by Google to target newer API levels which trigger the new exception.

Well, maybe I misunderstand you when you say "replicate the crash with 3.8.0". I'd hope that there is no crash when switching to 3.8.0. Is this just throwing another exception?

The `com.dsi.ant.plugins.utility.log.LogAnt` shouldn't access preferences with "MODE_WORLD_READABLE". Has this been changed?

**Comment by harrychin** [comment]
> Well, maybe I misunderstand you when you say "replicate the crash with 3.8.0". I'd hope that there is no crash when switching to 3.8.0. Is this just throwing another exception?
> 
> The `com.dsi.ant.plugins.utility.log.LogAnt` shouldn't access preferences with "MODE_WORLD_READABLE". Has this been changed?

Correct, I was referencing the "crash" inside this issue, but it's just the `SecurityException` being left uncaught.

As you've intuited, `com.dsi.ant.plugins.utility.log.LogAnt` was using `MODE_WORLD_READABLE`, but that depreciation avoided linting/testing because Android references the API level for the package namespace from the AntPlusPluginService. It's only until after the service was updated, then the check for the `SecurityException` turned true for all apps which call into `bindAndRequest()`, which subsequently calls into `LogAnt`.

This version of the library no longer calls into a global shared preferences file for setting the debug level.

**Comment by marcoalt** [comment]
sorry for the dumb question, how do I use the .aar file? Is it possible to provide a .jar, so that we can simply replace the faulty library? Thank you for your help.

**Comment by iforpowell** [comment]
Just use it the same as a .jar.  Android studio and gradle will do the right thing.  I have this in my build.gradle dependencies section.

    implementation files('../libs/antpluginlib_3-8-0.aar')

If your not using Android studio or Gradle then time for some Googleing...
